### PR TITLE
OpenSSL integration in to KISSNET

### PR DESCRIPTION
**Pull request by cpz**
Can be improved but atm I'm using this version.

### CODE REVIEW

**Comment by cpz** [comment]
Notice, current version doesnt support multi-threading, It can be easily added by reading official OpenSSL examples for multi-threading(https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/crypto/threads/mttest.c)

**Comment by Ybalrid** [comment]
Hey, thanks for getting back to me on this. Will check that out this week-end 😄

File: kissnet.hpp [hunk 1]
```diff
@@ -583,6 +598,12 @@ namespace kissnet
 				type	  = SOCK_STREAM;
 				iprotocol = IPPROTO_TCP;
 			}
+			
+			else if constexpr (sock_proto == protocol::tcp_ssl)
```
Comment (Ybalrid) [inline kissnet.hpp]: It functionally doesn't change anything, but the code would be nicer without the duplication of the code between tcp and tcp_ssl.

I would propose to have one branch like this

```cpp
if constexpr(sock_proto == protocol::tcp || sock_proto == protocol::tcp_ssl)
```

File: kissnet.hpp [hunk 2]
```diff
@@ -762,6 +783,31 @@ namespace kissnet
...
+
+				pContext = SSL_CTX_new(ssl_method);
+				ssl = SSL_new(pContext);
+				if (!ssl)
+					return false;
+
+				if (!(static_cast<bool>(SSL_set_fd(ssl, sock))))
+					return false;
+
+				auto err = SSL_connect(ssl);
```
Comment (Ybalrid) [inline kissnet.hpp]: If we don't do anything with the `err` value beside returning false, maybe just do 

```cpp
if(SSL_connect(ssl) <= 0)
    return false;
```

File: kissnet.hpp [hunk 3]
```diff
@@ -807,8 +853,23 @@ namespace kissnet
 
 		void close()
 		{
-			if(!(sock == INVALID_SOCKET))
+			if (!(sock == INVALID_SOCKET))
```
Comment (Ybalrid) [inline kissnet.hpp]: This syntax is easier to read:
```cpp
if (sock != INVALID_SOCKET)
```

File: kissnet.hpp [hunk 4]
```diff
@@ -567,6 +578,10 @@ namespace kissnet
 
 		///operatic-system type for a socket object
 		SOCKET sock;
+#ifdef KISSNET_USE_OPENSSL
+		SSL *ssl;
```
Comment (Ybalrid) [inline kissnet.hpp]: Are these pointer taking ownership of the SSL and SSL_CTX objects here? If so, they **need** to be properly handled in the move constructor and the destructor of the class.

Best would be to have them as std::unique_ptr<>. But at the very least we need to make sure to copy and reset these pointers in move operations of the socket class (i.e. `socket(socket&&)` and `socket& operator=(socket&&)`

File: kissnet.hpp (hunk 1 above)
Comment (cpz) [reply kissnet.hpp]: Done!

File: kissnet.hpp (hunk 2 above)
Comment (cpz) [reply kissnet.hpp]: Done!

File: kissnet.hpp (hunk 3 above)
Comment (cpz) [reply kissnet.hpp]: Done!

File: kissnet.hpp (hunk 4 above)
Comment (cpz) [reply kissnet.hpp]: I didn't figured how to move them properly in classes and etc, but will look at this little bit later.

Comment (Ybalrid) [reply kissnet.hpp]: The gist of it is, in the move constructor/move assignment operator you want 2 things to occur : 

1) The "moved to" object get the value of these pointers
2) The "moved from" object get these pointers set back to `nullptr`

Move semantics in C++ are recent (C++11 and up) and are genuinely a bit confusing, especially the fact that they use these "r-value references" (you know `type&&`).

I know many C++ programmer that learned the old school stuff and did not refresh to the new standards thoroughly struggle with that specific thing. If you want to read more about it you should google for "C++11 move semantics"

Comment (cpz) [reply kissnet.hpp]: I've did research and seems like unique_ptr are not required for SSL pointers. I can be wrong but its what I got.

Did std::move for SSL pointers and better way initialization of OpenSSL.

Comment (Ybalrid) [reply kissnet.hpp]: I'll take a look but `std::move()` should not be required. A pointer's just a number.

Comment (Ybalrid) [reply kissnet.hpp]: Actually, I used std::move everywhere there... It's technically not wrong so whatever. I'll fix that later.

I need to //TODO some automated testing for this kind of stuff to be honest, but that's a subject *for another day* 🤔

**Comment by Ybalrid** [comment]
Looking good! Gonna play a bit with your new code and merge!

File: kissnet.hpp (hunk 4 above)
Comment (cpz) [reply kissnet.hpp]: Soon I'm going to learn about catch library and can try to integrate it in kissnet if need.

File: kissnet.hpp [hunk 5]
```diff
@@ -119,6 +120,9 @@
 #include <string>
 #include <utility>
 
+#include <openssl\ssl.h>
+#include <openssl\err.h>
```
Comment (Ybalrid) [inline kissnet.hpp]: Paths with anti-slashes only works with Microsoft Windows. That's a big nono. Also these includes needs to be in the `#ifdef` for openssl!

Comment (Ybalrid) [reply kissnet.hpp]: Actually in latest version it's in the #ifdef... But still, don't use antislashes, ever! 😉

**Comment by Ybalrid** [comment]
`TLSv1_2_client_method()` is deprecated in latests OpenSSL versions. I feel like we should build this against the latest OpenSSL version available (1.1.1?). I *know* that you are explicitly checking against that version of the library. What's the issue here exactly?

**Comment by cpz** [comment]
> `TLSv1_2_client_method()` is deprecated in latests OpenSSL versions. I feel like we should build this against the latest OpenSSL version available (1.1.1?). I _know_ that you are explicitly checking against that version of the library. What's the issue here exactly?

There a limit usage in mine situation. I can use only 1.0.2. I can make version which supports 1.1.1

**Comment by Ybalrid** [comment]
I need to check about API differences with OpenSSL. 

Beside *one warning*, (and with with the `\` replaced by `/` in the `#include` for SSL) your code now works on my ArchLinux dev environment that is using `OpenSSL 1.1.1d` so we're quite good here already ;-)

I have been able to do a "GET /" on an HTTPS server just by declaring an SSL socket and doing `connect`/`send`/`recv`. So that works as advertised 😉

**Comment by cpz** [comment]
> I need to check about API differences with OpenSSL.
> 
> Beside _one warning_, (and with with the `\` replaced by `/` in the `#include` for SSL) your code now works on my ArchLinux dev environment that is using `OpenSSL 1.1.1d` so we're quite good here already 
> 
> I have been able to do a "GET /" on an HTTPS server just by declaring an SSL socket and doing `connect`/`send`/`recv`. So that works as advertised

If OpenSSL version > 1.1.0
then for initialize we should use:
```C++
OPENSSL_init_ssl(
        OPENSSL_INIT_LOAD_SSL_STRINGS | OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);

So Initialize should be something like this:
```

**Comment by cpz** [comment]
Oh, my bad, wrong button. Wanted to post comment.

If OpenSSL > 1.1.0 seems like we should do this:
```C++
	#ifdef KISSNET_USE_OPENSSL
	class Initialize_SSL
	{
	public:
		Initialize_SSL()
		{
            #if OPENSSL_VERSION_NUMBER < 0x10100000L
			SSL_load_error_strings();
			SSL_library_init();
			SSLeay_add_ssl_algorithms();
			#else
			OPENSSL_init_ssl(OPENSSL_INIT_LOAD_SSL_STRINGS | OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);
			#endif
		}

		~Initialize_SSL()
		{
			#if OPENSSL_VERSION_NUMBER < 0x10100000L
			ERR_free_strings();
			EVP_cleanup();
			#endif
		}
	};

	static Initialize_SSL InitializeSSL;
	#endif
```

If I get right documentation of OpenSSL

**Comment by Ybalrid** [comment]
I'm all for the static version check in the header with `#if`. If you cannot test on your dev' environment because you are stuck on 1.0.2 that's fine, I'll do my testing against the latest versions

**Comment by Ybalrid** [comment]
It's 1am here. I'll look at all of this tomorrow in the evening (I'm in CET timezone). Do the SSL version fix and the #include fix on your fork and I'll check that out.

**Comment by Ybalrid** [comment]
In the current state, this code *works*. There's no reason *not* to merge it to master. 

It will generate some warnings on newer versions of OpenSSL, but these changes check the "Get SSL working" so that's good.

Many thanks!

fix #8

**Pull request merged**